### PR TITLE
Earcut.js

### LINKED ISSUE #2788: Large polygon triangulation produces incorrect results
Via [the forum](https://groups.google.com/d/msg/cesium-dev/2QcY4BtF4zo/1eyk1KkWsEEJ)

The two polygons below render fine.  But if you create a single polygon where the second one is a hole in the first, you get totally incorrect results.  Note, the 179.9999 is a workaround for the fact that we still fill large polygons, which I believe we have a bug on still.

```javascript
var viewer = new Cesium.Viewer('cesiumContainer');

viewer.entities.add({
	polygon: {
		hierarchy:
        new Cesium.PolygonHierarchy(Cesium.Cartesian3.fromDegreesArray([
				-179.9999, 80,
				-179.9999, -80,
				0.0, -80,
				0.0, 80])),
		material: Cesium.Color.WHITE.withAlpha(0.5)
	}
});

viewer.entities.add({
	polygon: {
		hierarchy:
        new Cesium.PolygonHierarchy(Cesium.Cartesian3.fromDegreesArray([
						-97.0, 31.0,
						-97.0, 39.0,
						-87.0, 39.0,
						-87.0, 31.0
					])),
		material: Cesium.Color.BLUE.withAlpha(0.5)
	}
});
```

### PR DESCRIPTION
**Pull request by bagnell**
Uses earcut.js for triangulation. Fixes #2788. Fixes some of the files uploaded to #3044, but the ones that still don't work look better. The files in #1209 work. The problem with the US states happens because of the intersecting edges.

TODO:
- [x] Remove `eliminateHoles` code
- [x] Remove `triangulate` code
- [x] Update tests
- [x] Profile
- [x] Clean up
- [x] Update LICENSE.md
- [x] Update CHANGES.md

### CODE REVIEW

**Comment by pjcozzi** [comment]
All looks good so far!

File: Source/Core/PolygonPipeline.js [hunk 1]
```diff
@@ -743,33 +82,22 @@ define([
...
      */
-    PolygonPipeline.triangulate = function(positions) {
+    PolygonPipeline.triangulate = function(positions, holes) {
         //>>includeStart('debug', pragmas.debug);
         if (!defined(positions)) {
             throw new DeveloperError('positions is required.');
         }
-        if (positions.length < 3) {
-            throw new DeveloperError('At least three positions are required.');
+        if (!defined(holes)) {
```
Comment (mramato) [inline Source/Core/PolygonPipeline.js]: Why is this required, it doesn't look like `earcut` requires it.

Comment (bagnell) [inline Source/Core/PolygonPipeline.js]: You're right, thanks.

**Comment by bagnell** [comment]
@pjcozzi This is ready for another look.

**Comment by pjcozzi** [comment]
All looks good to me.

@mramato can you test the OSM NYC dataset offline?

@bagnell are there any other datasets you want to test with?

**Comment by pjcozzi** [comment]
Do you have a representative total performance improvement for a large dataset?

**Comment by bagnell** [comment]
@pjcozzi Triangulation when loading all of the country borders takes 305.322ms on average for the whole dataset in the branch and 641.598ms in master.

**Comment by pjcozzi** [comment]
Sounds great.  Can you add this example performance improvement to CHANGES.md?

Is anything else needed before we can merge this?

**Comment by mramato** [comment]
FYI, I ran this on the full NYC set I was working with last week and it fixed tons of problems that we were having with the old triangulation code.  So this is a big :+1: from me.

**Comment by bagnell** [comment]
CHANGES.md was updated. I think this is good to go.

**Comment by pjcozzi** [comment]
@bagnell given that this will be of interest to many users, can you please email the forum to let them know this is coming in Cesium 1.23 on July 1?

File: CHANGES.md [hunk 2]
```diff
@@ -5,6 +5,9 @@ Change Log
 
 * Add a `rotatable2D` option to to `Scene`, `CesiumWidget` and `Viewer` to enable a rotatable map in 2D. [#3897](https://github.com/AnalyticalGraphicsInc/cesium/issues/3897)
 * `Camera.setView` and `Camera.flyTo` will now use the `orientation.heading` parameter in 2D if the map is rotatable.
+* Added `packArray` and `unpackArray` functions to `Cartesian2`, `Cartesian3`, and `Cartesian4`.
+* Fix some large polygon triangulations. [#2788](https://github.com/AnalyticalGraphicsInc/cesium/issues/2788)
+* Improved performance and accuracy of polygon triangulation by using the [earcut](https://github.com/mapbox/earcut) library. Loading a GeoJSON with polygons for each country was ~50% faster.
```
Comment (pjcozzi) [inline CHANGES.md]: Do you mean 2x faster, i.e., 50% less time?

Comment (bagnell) [inline CHANGES.md]: Yup. I'll fix it in master.

**Comment by mourner** [comment]
Awesome work, happy to see it used in Cesium!

**Pull request merged**